File: java/library/src/main/java/com/v2soft/styxlib/messages/base/StyxTMessage.java
package com.v2soft.styxlib.messages.base;

import com.v2soft.styxlib.exceptions.StyxErrorMessageException;
import com.v2soft.styxlib.exceptions.StyxException;
import com.v2soft.styxlib.exceptions.StyxWrongMessageException;
import com.v2soft.styxlib.messages.base.enums.MessageType;

import java.util.concurrent.TimeoutException;

public class StyxTMessage extends StyxMessage {
    private StyxMessage mAnswer;
    private MessageType mRequiredAnswerType;

    public StyxTMessage(MessageType type, MessageType answer) {
        super(type, StyxMessage.NOTAG);
        mRequiredAnswerType = answer;
    }

    public StyxMessage getAnswer() {
        return mAnswer;
    }

    public void setAnswer(StyxMessage answer)
            throws StyxException {
        synchronized (this) {
            if (!checkAnswer(answer))
                throw new StyxWrongMessageException(answer, mRequiredAnswerType);
            mAnswer = answer;
            notifyAll();
        }
    }

    public synchronized StyxMessage waitForAnswer(long timeout) throws InterruptedException, TimeoutException,
            StyxErrorMessageException {
        if ( mAnswer == null)
            wait(timeout);
        if ( mAnswer == null )
            throw new TimeoutException("Don't receive answer for "+this.toString());
        if (mAnswer.getType() == MessageType.Rerror) {
            StyxErrorMessageException.checkException(mAnswer);
        }
        return mAnswer;
    }

    protected boolean checkAnswer(StyxMessage answer) {
        final MessageType received = answer.getType();
        return (mRequiredAnswerType == received || received == MessageType.Rerror);
    }
}

File: native/V2StyxLib/src/messages/base/StyxTMessage.cpp
/*
 * StyxTMessage.cpp
 *
 *  Created on: Apr 06, 2018
 *      Author: vschryabets@gmail.com
 */

#include "messages/base/StyxTMessage.h"
#include "exceptions/StyxWrongMessageException.h"

StyxTMessage::StyxTMessage(MessageTypeEnum type, MessageTypeEnum answer)
	: StyxMessage(type, StyxMessage::NOTAG),
	  mRequiredAnswerType(answer), mWaitSyncObject(NULL), mAnswer(NULL)
{
}

StyxTMessage::~StyxTMessage() {
}

StyxMessage* StyxTMessage::getAnswer() {
	return mAnswer;
}

void StyxTMessage::setAnswer(StyxMessage* answer) throw(StyxException) {
	if (!checkAnswer(answer))
		throw StyxWrongMessageException(answer, mRequiredAnswerType);
#warning set answer should be after mutex locked
	mAnswer = answer;
	if (mWaitSyncObject != NULL) {
		mWaitSyncObject->notifyAll();
	}
}

StyxMessage* StyxTMessage::waitForAnswer() 
	throw(StyxErrorMessageException) 
{
	mWaitSyncObject->waitForNotify(&mAnswer);
	if ( mAnswer == NULL )
		throw StyxException("Don't receive answer for %s", this->toString().c_str());
	if (mAnswer->getType() == Rerror) {
		StyxErrorMessageException::checkException(mAnswer);
	}
	return mAnswer;
}

#warning do we need this method?
bool StyxTMessage::checkAnswer(StyxMessage* answer) 
{
	MessageTypeEnum received = answer->getType();
	return (mRequiredAnswerType == received || received == Rerror);
}

void StyxTMessage::setSyncObject(SyncObject* syncObject)
{
	mWaitSyncObject = syncObject;
}
